stores/filestore.go

```diff
@@ -665,6 +715,286 @@ func (w *bufferedWriter) checkShrinkRequest() {
+// before the file handle is closed.
+func (fm *filesManager) closeLockedOrOpenedFile(file *file) error {
+	// Check first locked files
+	if !atomic.CompareAndSwapInt32(&file.state, fileInUse, fileClosing) {
+		// then opened but unlocked files
+		if !atomic.CompareAndSwapInt32(&file.state, fileOpened, fileClosing) {
+			return nil
+		}
+	}
+	fm.Lock()
+	err := fm._doClose(file)
+	fm.Unlock()
+	return err
+}
+
+// _doClose closes the file handle, setting it to nil and switching state to `fileClosed`.
+// If a `beforeClose` callback was registered on file creation, it is invoked
+// before the file handler is actually closed.
+// Lock is required on entry.
+func (fm *filesManager) _doClose(file *file) error {
+	var err error
+	if file.beforeClose != nil {
+		err = file.beforeClose()
+	}
+	util.CloseFile(err, file.handle)
+	// Regardless of error, we need to change the state to closed.
+	file.handle = nil
+	atomic.StoreInt32(&file.state, fileClosed)
+	fm.opened--
+	return err
+}
+
+// lockFile locks the given file.
+// If the file was already opened, the boolean returned is true,
+// otherwise, the file is opened and the call returns false.
+func (fm *filesManager) lockFile(file *file) (bool, error) {
+	if atomic.CompareAndSwapInt32(&file.state, fileOpened, fileInUse) {
+		return true, nil
+	}
+	return false, fm._openFile(file)
+}
+
+// lockFileIfOpened is like lockFile but returns true only if the
+// file is already opened, false otherwise (and the file remain closed).
+func (fm *filesManager) lockFileIfOpened(file *file) bool {
+	return atomic.CompareAndSwapInt32(&file.state, fileOpened, fileInUse)
+}
+
+// unlockFile unlocks the file if currently locked, otherwise panic.
+func (fm *filesManager) unlockFile(file *file) {
+	if !atomic.CompareAndSwapInt32(&file.state, fileInUse, fileOpened) {
+		panic(fmt.Errorf("failed to switch state from fileInUse to fileOpened for file %q, state=%v",
+			file.name, file.state))
+	}
+}
+
+// _trySwitchState attempts to switch an initial state of `fileOpened`
+// or `fileClosed` to the given newState. If it can't it will return an
+// error, otherwise, returned a boolean to indicate if the initial state
+// was `fileOpened`.
+func (fm *filesManager) _trySwitchState(file *file, newState int32) (bool, error) {
+	wasOpened := false
+	wasClosed := false
+	for i := 0; i < 10000; i++ {
+		if atomic.CompareAndSwapInt32(&file.state, fileOpened, newState) {
+			wasOpened = true
+			break
+		}
+		if atomic.CompareAndSwapInt32(&file.state, fileClosed, newState) {
+			wasClosed = true
+			break
+		}
+		if i%1000 == 1 {
+			time.Sleep(time.Millisecond)
+		}
+	}
+	if !wasOpened && !wasClosed {
+		return false, fmt.Errorf("file %q is still probably locked", file.name)
+	}
+	return wasOpened, nil
+}
+
+// remove a file from the list of files. The initial state must be either `fileOpened`
+// or `fileClosed`. This call will loop until it can switch the file's state from
```

Comment:
This function is calling the function above. So yes, it will loop for a certain number of attempts if the file is not in the state it should be.